File: DraftRightWindows/DraftRightWindows/Services/FeedbackService.cs
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;

namespace DraftRightWindows.Services;

/// <summary>
/// Posts feature-request feedback to the DraftRight backend's <c>POST /feedback</c>
/// endpoint as <c>application/json</c>. Anonymous and authenticated calls are
/// both supported.
///
/// Mirrors <see cref="BugReportService"/> for URL/token resolution and
/// <see cref="SubmitResult"/> shape.
/// </summary>
public static class FeedbackService
{
    private static readonly HttpClient _http = new() { Timeout = TimeSpan.FromSeconds(30) };
    private const string DefaultBackendUrl = "https://api.draftright.info";

    /// <summary>
    /// Result of a feedback submission. <see cref="Success"/> indicates HTTP 2xx;
    /// <see cref="ErrorMessage"/> carries the failure reason otherwise.
    /// </summary>
    public sealed record SubmitResult(bool Success, string? Id = null, string? ErrorMessage = null);

    /// <summary>
    /// POSTs a feature request to <c>/feedback</c>.
    /// </summary>
    /// <param name="title">Short title for the request (max 80 chars, enforced by caller).</param>
    /// <param name="targetPlatform">One of: playground|mobile|windows|mac|linux.</param>
    /// <param name="description">Longer description of the feature.</param>
    /// <param name="userEmail">Optional. Sent only when <paramref name="authToken"/> is absent.</param>
    /// <param name="authToken">Optional. Bearer JWT — when present server attaches user_id.</param>
    /// <param name="backendUrlOverride">Optional. Override base URL (tests / dev). Defaults to settings.</param>
    public static async Task<SubmitResult> SubmitAsync(
        string title,
        string targetPlatform,
        string description,
        string? userEmail = null,
        string? authToken = null,
        string? backendUrlOverride = null)
    {
        if (string.IsNullOrWhiteSpace(title))
            return new SubmitResult(false, ErrorMessage: "Title is required.");

        if (string.IsNullOrWhiteSpace(description))
            return new SubmitResult(false, ErrorMessage: "Description is required.");

        // Resolve base URL: explicit override > current Settings > hardcoded default.
        // App.Settings is null in unit/standalone scenarios, so guard before reaching it.
        var baseUrl = backendUrlOverride;
        if (string.IsNullOrWhiteSpace(baseUrl))
        {
            try { baseUrl = App.Settings?.BackendUrl; } catch { baseUrl = null; }
        }
        if (string.IsNullOrWhiteSpace(baseUrl)) baseUrl = DefaultBackendUrl;
        baseUrl = baseUrl!.TrimEnd('/');

        try
        {
            var payload = new Dictionary<string, object?>
            {
                ["kind"] = "feature",
                ["title"] = title.Trim(),
                ["target_platform"] = targetPlatform,
                ["description"] = description.Trim(),
                ["source"] = "windows-app",
            };

            if (string.IsNullOrWhiteSpace(authToken) && !string.IsNullOrWhiteSpace(userEmail))
                payload["user_email"] = userEmail!.Trim();

            using var request = new HttpRequestMessage(HttpMethod.Post, $"{baseUrl}/feedback")
            {
                Content = new StringContent(
                    JsonSerializer.Serialize(payload),
                    Encoding.UTF8,
                    "application/json"),
            };

            if (!string.IsNullOrEmpty(authToken))
                request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", authToken);

            using var response = await _http.SendAsync(request).ConfigureAwait(false);
            var body = await response.Content.ReadAsStringAsync().ConfigureAwait(false);

            if (!response.IsSuccessStatusCode)
            {
                string detail;
                try
                {
                    using var doc = JsonDocument.Parse(body);
                    if (doc.RootElement.TryGetProperty("error", out var err))
                        detail = err.GetString() ?? body;
                    else if (doc.RootElement.TryGetProperty("message", out var msg))
                        detail = msg.GetString() ?? body;
                    else
                        detail = body;
                }
                catch
                {
                    detail = body;
                }
                return new SubmitResult(
                    false,
                    ErrorMessage: $"Server {(int)response.StatusCode}: {detail}");
            }

            string? id = null;
            try
            {
                using var doc = JsonDocument.Parse(body);
                if (doc.RootElement.TryGetProperty("id", out var idProp))
                    id = idProp.GetString();
            }
            catch
            {
                // 2xx but unparseable body — still treat as success.
            }

            return new SubmitResult(true, Id: id);
        }
        catch (TaskCanceledException)
        {
            return new SubmitResult(false, ErrorMessage: "Request timed out. Check your connection and retry.");
        }
        catch (Exception ex)
        {
            return new SubmitResult(false, ErrorMessage: ex.Message);
        }
    }
}

File: DraftRight/Services/FeedbackService.swift
import Foundation

/// Posts feature requests to the backend `POST /feedback` endpoint
/// (JSON body; no screenshot). Mirrors `BugReportService` but for feature
/// requests rather than bug reports.
enum FeedbackService {
    static let source = "macos-app"

    enum FeedbackError: LocalizedError {
        case invalidURL
        case httpError(Int, String)
        case emptyResponse
        case network(Error)

        var errorDescription: String? {
            switch self {
            case .invalidURL: return "Invalid backend URL."
            case .httpError(let code, let body): return "HTTP \(code): \(body)"
            case .emptyResponse: return "No response from server."
            case .network(let err): return err.localizedDescription
            }
        }
    }

    struct CreateResponse: Decodable { let id: String }

    /// Submit a feature request.
    /// - Parameters:
    ///   - title: One-line summary of the requested feature.
    ///   - targetPlatform: One of playground|mobile|windows|mac|linux.
    ///   - description: Detailed description.
    ///   - userEmail: Optional email for anonymous flow (ignored when `authToken` present).
    ///   - authToken: Optional Bearer JWT — pass non-empty access token when signed in.
    ///   - backendUrl: Base URL e.g. `https://api.draftright.info`.
    /// - Returns: Server-assigned record ID (may be empty string if server omits it).
    static func submitFeatureRequest(
        title: String,
        targetPlatform: String,
        description: String,
        userEmail: String?,
        authToken: String?,
        backendUrl: String
    ) async throws -> String {
        let base = backendUrl.strippingTrailingSlash
        guard let url = URL(string: "\(base)/feedback") else {
            throw FeedbackError.invalidURL
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.timeoutInterval = 30
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let token = authToken, !token.isEmpty {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        var body: [String: Any] = [
            "kind": "feature",
            "title": title.trimmingCharacters(in: .whitespacesAndNewlines),
            "target_platform": targetPlatform,
            "description": description.trimmingCharacters(in: .whitespacesAndNewlines),
            "source": source,
        ]
        // Only include email in anonymous flow — authenticated requests bind to the account
        if (authToken ?? "").isEmpty,
           let email = userEmail?.trimmingCharacters(in: .whitespaces),
           !email.isEmpty {
            body["user_email"] = email
        }

        let data: Data
        let response: URLResponse
        do {
            request.httpBody = try JSONSerialization.data(withJSONObject: body)
            (data, response) = try await URLSession.shared.data(for: request)
        } catch let err as FeedbackError {
            throw err
        } catch {
            throw FeedbackError.network(error)
        }

        guard let http = response as? HTTPURLResponse else {
            throw FeedbackError.emptyResponse
        }
        if http.statusCode >= 400 {
            let bodyText = String(data: data, encoding: .utf8) ?? ""
            throw FeedbackError.httpError(http.statusCode, bodyText)
        }

        return (try? JSONDecoder().decode(CreateResponse.self, from: data))?.id ?? ""
    }
}
